cpp/src/c_api/array.cpp

```diff
@@ -0,0 +1,183 @@
+
+#include <raft/handle.hpp>
+
+#include <rmm/device_buffer.hpp>
+
+namespace c_api {
+
+typedef struct {
+  rmm::device_buffer* data_;
+  size_t size_;
+  size_t nbytes_;
+  data_type_id_t type_;
+} cugraph_type_erased_device_array_t;
+
+typedef struct {
+  std::byte* data_;
+  size_t size_;
+  size_t nbytes_;
+  data_type_id_t type_;
+} cugraph_type_erased_host_array_t;
+
+}  // namespace c_api
+
+extern "C" cugraph_error_t cugraph_type_erased_device_array_create(
+  const cugraph_handle_t* handle,
+  data_type_id_t dtype,
+  size_t n_elems,
+  cugraph_type_erased_device_array_t** array)
+{
+  *array = nullptr;
+
+  try {
+    raft::handle_t const* raft_handle = reinterpret_cast<raft::handle_t const*>(handle);
+
+    if (!raft_handle) return CUGRAPH_INVALID_HANDLE;
+
+    size_t nbytes = n_elems * (::data_type_sz[dtype]);
+
+    c_api::cugraph_type_erased_device_array_t* ret_value =
+      new c_api::cugraph_type_erased_device_array_t{
+        new rmm::device_buffer(nbytes, raft_handle->get_stream()), n_elems, nbytes, dtype};
+
+    *array = reinterpret_cast<cugraph_type_erased_device_array_t*>(ret_value);
+    return CUGRAPH_SUCCESS;
+  } catch (...) {
+    return CUGRAPH_UNKNOWN_ERROR;
+  }
+}
+
+extern "C" void cugraph_type_erased_device_array_free(cugraph_type_erased_device_array_t* p)
+{
+  auto internal_pointer = reinterpret_cast<c_api::cugraph_type_erased_device_array_t*>(p);
+  delete internal_pointer->data_;
+  delete internal_pointer;
+}
+
+extern "C" size_t cugraph_type_erased_device_array_size(const cugraph_type_erased_device_array_t* p)
+{
+  auto internal_pointer = reinterpret_cast<c_api::cugraph_type_erased_device_array_t const*>(p);
+  return internal_pointer->size_;
+}
+
+extern "C" data_type_id_t cugraph_type_erased_device_array_type(
+  const cugraph_type_erased_device_array_t* p)
+{
+  auto internal_pointer = reinterpret_cast<c_api::cugraph_type_erased_device_array_t const*>(p);
+  return internal_pointer->type_;
+}
+
+extern "C" void* cugraph_type_erased_device_array_pointer(
+  const cugraph_type_erased_device_array_t* p)
+{
+  auto internal_pointer = reinterpret_cast<c_api::cugraph_type_erased_device_array_t const*>(p);
+  return internal_pointer->data_->data();
+}
+
+extern "C" cugraph_error_t cugraph_type_erased_host_array_create(
+  const cugraph_handle_t* handle,
+  data_type_id_t dtype,
+  size_t n_elems,
+  cugraph_type_erased_host_array_t** array)
+{
+  *array = nullptr;
+
+  try {
+    raft::handle_t const* raft_handle = reinterpret_cast<raft::handle_t const*>(handle);
+
+    if (!raft_handle) return CUGRAPH_INVALID_HANDLE;
+
+    size_t nbytes = n_elems * (::data_type_sz[dtype]);
+
+    c_api::cugraph_type_erased_host_array_t* ret_value =
+      new c_api::cugraph_type_erased_host_array_t{new std::byte[nbytes], n_elems, nbytes, dtype};
+
+    *array = reinterpret_cast<cugraph_type_erased_host_array_t*>(ret_value);
+    return CUGRAPH_SUCCESS;
+  } catch (...) {
+    return CUGRAPH_UNKNOWN_ERROR;
```

Comment:
Should there be a `FIXME` here to add an exception handler that checks for more error situations and sets the code with better information (eg. ALLOC error, etc.).

Reply:
Looks like my main concern was handled by setting the exception message on the new error type.